rust/src/mqtt/parser.rs

```diff
@@ -0,0 +1,631 @@
+           >>  (
+                 (name, value)
+               )
+       ));
+
+// MESSAGE COMPONENTS
+
+named!(#[inline], pub parse_property<MQTTProperty>,
+       do_parse!(
+           identifier: parse_mqtt_variable_integer
+           >> value: call!(parse_qualified_property, identifier)
+           >>  (
+                 value
+               )
+       ));
+
+#[inline]
+fn parse_properties(input: &[u8], precond: bool) -> IResult<&[u8], Option<Vec<MQTTProperty>>> {
+    // do not try to parse anything when precondition is not met
+    if !precond {
+        return Ok((input, None));
+    }
+    // parse properties length
+    match parse_mqtt_variable_integer(input) {
+        Ok((rem, mut proplen)) => {
+            if proplen == 0 {
+                // no properties
+                return Ok((rem, None));
+            }
+            // parse properties
+            let mut props = Vec::<MQTTProperty>::new();
+            let mut newrem = rem;
+            while proplen > 0 {
+                match parse_property(newrem) {
+                    Ok((rem, val)) => {
+                        props.push(val);
+                        let curparselen = (newrem.len() - rem.len()) as u32;
+                        proplen -= curparselen;
+                        newrem = &rem;
+                    }
+                    Err(e) => return Err(e),
+                }
+            }
+            return Ok((newrem, Some(props)));
+        }
+        Err(e) => return Err(e),
+    }
+}
+
+#[inline]
+fn parse_fixed_header_flags(i: &[u8]) -> IResult<&[u8], (u8, u8, u8, u8)> {
+    bits!(
+        i,
+        tuple!(
+            take_bits!(4u8),
+            take_bits!(1u8),
+            take_bits!(2u8),
+            take_bits!(1u8)
+        )
+    )
+}
+
+named!(#[inline], pub parse_fixed_header<FixedHeader>,
+       do_parse!(
+           flags: parse_fixed_header_flags
+           >> remaining_length: parse_mqtt_variable_integer
+           >>  (
+                 FixedHeader {
+                   message_type: flags.0 as u8,
+                   dup_flag: flags.1 != 0,
+                   qos_level: flags.2 as u8,
+                   retain: flags.3 != 0,
+                   remaining_length: remaining_length,
+                 }
+               )
+       ));
+
+#[inline]
+fn parse_connect_variable_flags(i: &[u8]) -> IResult<&[u8], (u8, u8, u8, u8, u8, u8, u8)> {
+    bits!(
+        i,
+        tuple!(
+            take_bits!(1u8),
+            take_bits!(1u8),
+            take_bits!(1u8),
+            take_bits!(2u8),
+            take_bits!(1u8),
+            take_bits!(1u8),
+            take_bits!(1u8)
+        )
+    )
+}
+
+named!(#[inline], pub parse_connect<MQTTConnectData>,
+       do_parse!(
+           protocol_string: parse_mqtt_string
+           >> protocol_version: be_u8
+           >> flags: parse_connect_variable_flags
+           >> keepalive: be_u16
+           >> properties: call!(parse_properties, protocol_version == 5)
```

Comment:
Why not use `cond!` ?

Reply:
I would need to dive deeper and try to get my head around this again, but I think it was because of the variety here and the fact that `cond!` would already consume something that would need to be part of another parsed field in case the protocol version is not 5 (as there is no space for properties in that case). I definitely remember I tried various combinations involving `cond!` and it was more complicated that doing it like this. If you have a better solution, please feel free to propose it.

Reply:
Keep the working version, thanks